lib/memo_wise/internal_api.rb

```diff
@@ -10,15 +10,32 @@ class InternalAPI
     #
     # @return [Object] the passed-in obj
     def self.create_memo_wise_state!(obj)
-      # `@_memo_wise` stores memoized results of method calls. For performance
-      # reasons, the structure differs for different types of methods. It looks
-      # like:
+      # `@_memo_wise` and `@_memo_wise_single_argument` store memoized results
+      # of method calls. For performance reasons, the structure differs for
+      # different types of methods.
+      #
+      # `@_memo_wise` looks like:
       #   {
       #     no_args_method_name: :memoized_result,
-      #     single_arg_method_name: { arg1 => :memoized_result, ... },
       #     [:multi_arg_method_name, arg1, arg2].hash => :memoized_result
       #   }
+      #
+      # `@_memo_wise_single_argument` looks like:
+      #   [
+      #     { arg1 => :memoized_result, ... }, # For method 1
+      #     { arg1 => :memoized_result, ... }, # For method 2
+      #   ]
+      # This is essentially a faster alternative to:
+      #   {
+      #     single_arg_method_name: { arg1 => :memoized_result, ... }
+      #   }
+      # because we can give each single-argument method its own array index at
+      # load time and perform that array lookup more quickly than a hash lookup
+      # by method name.
       obj.instance_variable_set(:@_memo_wise, {}) unless obj.instance_variables.include?(:@_memo_wise)
+      unless obj.instance_variables.include?(:@_memo_wise_single_argument)
```

Comment:
I'd rather not do that now since at this point that hash also contains 0-arity methods, and I think it's possible that this optimization won't end up being faster for 0-arity methods because of the sentinel issue. Thoughts?

Reply:
That makes sense to me